File: desktop/src/shared/ui/markdown/pathLinks.ts
/**
 * Shape gate and IPC wrapper for local file paths written as inline code in a
 * message.
 *
 * Agents post report paths as text (`audit/verify/report.md`,
 * `buzz/approvals/item-7.html`). This module decides which inline-code tokens
 * *look* like a path and carries the click/hover resolution to the native
 * `resolve_path_link` command, which is the only place a filesystem is
 * touched. Nothing here reads a file, and nothing is resolved while a channel
 * renders — a token stays plain text until a hover or a click asks.
 *
 * The gate here is a cheap pre-filter, not the security boundary. Root
 * containment, symlink escape and the regular-file check all live in Rust
 * (`commands/path_links.rs`); this file only keeps obvious non-paths and
 * anything over the DTO's byte cap off the IPC channel entirely.
 *
 * Kept DOM-free so it is unit-testable without a webview.
 */

/**
 * Maximum UTF-8 byte length of a candidate the DTO will carry.
 *
 * Message content is relay-sourced and unbounded, so the cap is applied
 * before the string reaches the command. The native side enforces the same
 * number (`MAX_PATH_LINK_BYTES` in `commands/path_links.rs`) and refuses an
 * over-length candidate before any filesystem call — keep the two in sync.
 */
export const MAX_PATH_LINK_BYTES = 4096;

/**
 * Extensions that make a slash-free token a path candidate.
 *
 * A token with no slash is a path only when it names a document; a bare word
 * (`cargo`, `SIGKILL`) never qualifies, and neither does a slash-free script
 * or binary name — those are the tokens most likely to be prose or a command.
 *
 * The same list is the *native* allowlist for what a resolved link may be
 * (`OPENABLE_EXTENSIONS` in `commands/path_links.rs`): a token with a slash
 * still reaches the resolver, and the resolver refuses anything the OS
 * default handler would run rather than display. Keep the two in sync.
 */
const DOCUMENT_EXTENSIONS = [
  ".md",
  ".markdown",
  ".mdx",
  ".html",
  ".htm",
  ".pdf",
  ".csv",
  ".json",
  ".txt",
  ".log",
  ".yml",
  ".yaml",
  ".toml",
] as const;

/** Extensions the in-app markdown viewer renders. */
const MARKDOWN_EXTENSIONS = [".md", ".markdown", ".mdx"] as const;

/**
 * Scheme prefix marking a `?doc=` panel target as a local file rather than a
 * relay `/media/` URL. The panel reads local documents through
 * `read_path_link_markdown`, which re-resolves the path against the allowed
 * roots; a relay URL keeps the authenticated media fetch.
 */
const LOCAL_DOC_SCHEME = "buzz-local-file:";

/** What the native resolver decided a candidate is. */
export type PathLinkKind = "markdown" | "file";

/** A candidate that resolved to a real file inside an allowed root. */
export type PathLinkTarget = {
  /** Canonical absolute path, as resolved and contained natively. */
  path: string;
  /** Final path component, for the link title and the panel header. */
  filename: string;
  /**
   * `markdown` opens in the in-app viewer; `file` goes to the OS opener, and
   * the resolver has already proven it is an inert, non-executable document.
   */
  kind: PathLinkKind;
  /** Size of the resolved file in bytes. */
  sizeBytes: number;
};

/** The Tauri `invoke` seam, narrowed to what this module sends. */
export type PathLinkInvoke = (
  command: string,
  payload: Record<string, unknown>,
) => Promise<unknown>;

function utf8ByteLength(value: string): number {
  return new TextEncoder().encode(value).length;
}

/**
 * True when an inline-code token looks enough like a local path to be worth
 * one native resolution on hover or click.
 *
 * Refused outright, before any IPC: anything over the DTO byte cap, anything
 * carrying whitespace or a control character, a URL, a `~`-relative path, a
 * `..` traversal segment, and any bare word without a slash or a document
 * extension. A `true` here means "ask the resolver", never "this is a link" —
 * the resolver still has to find a regular file inside an allowed root.
 */
export function isPathLinkCandidate(text: string): boolean {
  // UTF-8 is never shorter than the UTF-16 code-unit count, so this rejects
  // an over-cap string without encoding it.
  if (text.length > MAX_PATH_LINK_BYTES) return false;
  if (text.length === 0) return false;
  if (utf8ByteLength(text) > MAX_PATH_LINK_BYTES) return false;

  // Whitespace (including the NBSP that pasted prose carries) and control
  // characters: a path token in a report never has them, and allowing them
  // would let one token smuggle a second argument to the opener.
  if (/[\s\p{Cc}]/u.test(text)) return false;

  // A URL is handled by the anchor renderer, never by the path resolver.
  if (text.includes("://")) return false;
  // No home expansion: `~` is not resolved anywhere in this path.
  if (text.startsWith("~")) return false;

  const segments = text.split("/");
  // `..` as a whole segment is a traversal attempt. `two..dots.md` is not.
  if (segments.some((segment) => segment === "..")) return false;

  // Must name something. `///` and `...` are not paths.
  if (!/[A-Za-z0-9]/.test(text)) return false;

  const lower = text.toLowerCase();
  const hasDocumentExtension = DOCUMENT_EXTENSIONS.some((extension) =>
    lower.endsWith(extension),
  );
  return text.includes("/") || hasDocumentExtension;
}

/** True when a resolved target opens in the in-app markdown viewer. */
export function isMarkdownPathLink(target: PathLinkTarget): boolean {
  const lower = target.filename.toLowerCase();
  return (
    target.kind === "markdown" &&
    MARKDOWN_EXTENSIONS.some((extension) => lower.endsWith(extension))
  );
}

/**
 * Validate the resolver's answer at the boundary.
 *
 * Returns `null` for `null`/`undefined` (the "not a link" answer) and for any
 * shape that is not a complete target, so a malformed reply renders as text
 * instead of a link pointing nowhere.
 */
export function parsePathLinkTarget(value: unknown): PathLinkTarget | null {
  if (value === null || typeof value !== "object") return null;
  const candidate = value as Partial<PathLinkTarget>;
  if (typeof candidate.path !== "string" || candidate.path.length === 0) {
    return null;
  }
  if (
    typeof candidate.filename !== "string" ||
    candidate.filename.length === 0
  ) {
    return null;
  }
  if (candidate.kind !== "markdown" && candidate.kind !== "file") return null;
  if (
    typeof candidate.sizeBytes !== "number" ||
    !Number.isFinite(candidate.sizeBytes)
  ) {
    return null;
  }
  return {
    path: candidate.path,
    filename: candidate.filename,
    kind: candidate.kind,
    sizeBytes: candidate.sizeBytes,
  };
}

/**
 * Resolve one candidate through the native command.
 *
 * Returns `null` without any IPC when the candidate fails the shape gate, and
 * `null` when the resolver reports "not a link" (a missing file, or a file
 * outside every allowed root). A resolver *error* — an over-length candidate
 * that got past the gate, an unreadable root — is re-thrown so the caller can
 * surface it rather than silently rendering plain text.
 */
export async function resolvePathLink(
  candidate: string,
  senderPubkey: string | null,
  invoke: PathLinkInvoke,
): Promise<PathLinkTarget | null> {
  if (!isPathLinkCandidate(candidate)) return null;
  const answer = await invoke("resolve_path_link", { candidate, senderPubkey });
  return parsePathLinkTarget(answer);
}

/** The `?doc=` panel target for a locally resolved markdown document. */
export function localMarkdownDocUrl(path: string): string {
  return `${LOCAL_DOC_SCHEME}${encodeURIComponent(path)}`;
}

/**
 * The local path behind a `?doc=` panel target, or `null` when the target is
 * a relay media URL (which the panel fetches over the authenticated path).
 */
export function parseLocalMarkdownDocUrl(url: string): string | null {
  if (!url.startsWith(LOCAL_DOC_SCHEME)) return null;
  try {
    const path = decodeURIComponent(url.slice(LOCAL_DOC_SCHEME.length));
    return path.length > 0 ? path : null;
  } catch {
    // A malformed percent-escape is not a path; the panel shows its load
    // error rather than sending the raw text to the filesystem.
    return null;
  }
}

File: desktop/src-tauri/src/commands/path_links.rs
//! Resolution of local file paths written as inline code in a message.
//!
//! Agents post report paths as text (`audit/verify/report.md`,
//! `buzz/approvals/item-7.html`). The webview decides which inline-code tokens
//! *look* like a path (`shared/ui/markdown/pathLinks.ts`); this module is the
//! only place that touches a filesystem, and it does so once per hover or
//! click, never while a channel renders.
//!
//! Containment contract: a candidate resolves only when its *canonical* form
//! is a regular file whose canonical path lies under a canonical allowed root.
//! Canonicalizing both sides is what makes `..` and an escaping symlink
//! equivalent to any other path outside the root — both land somewhere the
//! prefix check rejects. A candidate that names nothing is not an error: it is
//! "not a link", and the token stays plain text.
//!
//! Containment bounds *which* path a message can name. It says nothing about
//! what the handler does with that path, so this module bounds that too. The
//! OS default handler runs an executable, a `.command` script or a `.webloc`
//! shortcut, and `$HOME/projects` is full of all three — so a resolved file is
//! offered only when its extension is on [`OPENABLE_EXTENSIONS`] and it
//! carries no executable bit. Nothing here executes the file, and nothing
//! reaches the opener that the resolver has not just re-proven contained.

use std::io::Read;
use std::path::{Component, Path, PathBuf};

use serde::Serialize;
use tauri::AppHandle;
use tauri_plugin_opener::OpenerExt;

/// Maximum UTF-8 byte length of a candidate this module will look at.
///
/// Message content is relay-sourced and unbounded, so the cap is enforced at
/// the DTO — before any filesystem call — and mirrored in the webview's
/// `MAX_PATH_LINK_BYTES` so an over-length token never reaches IPC at all.
pub const MAX_PATH_LINK_BYTES: usize = 4096;

/// Maximum UTF-8 byte length of the sender pubkey carried alongside a
/// candidate. A pubkey is 64 hex characters; the cap leaves room for prefixed
/// forms without admitting an arbitrary relay-sourced string.
const MAX_SENDER_PUBKEY_BYTES: usize = 128;

/// Maximum size of a markdown document opened in the in-app viewer.
///
/// Matches the relay-attachment viewer's `MAX_MARKDOWN_DOC_BYTES`
/// (`media_download.rs`, `shared/ui/markdown/markdownDocFile.ts`) so a local
/// document and a shared one hit the same wall. A larger `.md` is still a
/// link — it opens with the OS handler instead of the panel.
pub const MAX_PATH_LINK_MARKDOWN_BYTES: u64 = 2 * 1024 * 1024;

/// Extensions the in-app markdown viewer renders.
const MARKDOWN_EXTENSIONS: [&str; 3] = ["md", "markdown", "mdx"];

/// Extensions a resolved path link may be offered for.
///
/// Mirrors `DOCUMENT_EXTENSIONS` in `shared/ui/markdown/pathLinks.ts`, and is
/// the guard that keeps a message from choosing what the OS default handler
/// runs. macOS `open` hands a file to LaunchServices, which *executes* a
/// `.command`, a `.sh` with an executable bit or any Mach-O binary, and
/// silently fetches the URL inside a `.webloc` or `.url`; Windows does the
/// same for `.exe`, `.cmd` and `.lnk`. `$HOME/projects` holds thousands of
/// those files, so an extension outside this inert-document list is "not a
/// link" rather than something a single click hands to the opener.
const OPENABLE_EXTENSIONS: [&str; 13] = [
    "md", "markdown", "mdx", "html", "htm", "pdf", "csv", "json", "txt", "log", "yml", "yaml",
    "toml",
];

/// How a resolved path link opens.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum PathLinkKind {
    /// A markdown document within the viewer cap: opens in the viewer panel.
    Markdown,
    /// Another inert document type: opens with the OS default handler.
    File,
}

/// A candidate that resolved to a real file inside an allowed root.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PathLinkTarget {
    /// Canonical absolute path. Always valid UTF-8: a canonical path that is
    /// not is refused, so this string re-resolves to the same file when the
    /// webview sends it back to open or read the document.
    pub path: String,
    /// Final path component, for the link title and the panel header.
    pub filename: String,
    /// Whether this opens in the viewer panel or with the OS handler.
    pub kind: PathLinkKind,
    /// Size of the resolved file in bytes.
    pub size_bytes: u64,
}

/// The working directory a message's sender writes into, when the desktop can
/// map that sender's pubkey to one.
///
/// There is no such mapping today. Every managed agent is spawned in the same
/// directory (`managed_agents::default_agent_workdir`, applied at
/// `managed_agents/runtime.rs:622`) and `ManagedAgentRecord` carries no
/// per-agent working directory, so a sender pubkey selects nothing and the
/// two shared roots below are the whole list. The seam is kept, and asserted
/// in `path_links_tests.rs`, so a future per-sender working directory has one
/// place to land and one test that records the change.
pub(super) fn sender_workdir_root(_sender_pubkey: Option<&str>) -> Option<PathBuf> {
    None
}

/// The nest (`~/.buzz`), which is the working directory every managed agent
/// is actually spawned in.
///
/// `managed_agents::default_agent_workdir` prefers the nest and falls back to
/// `$HOME`; only the nest half is a root here, because the fallback would
/// admit the whole home directory. A relative path an agent writes is
/// relative to this directory, so without it a relative candidate could only
/// ever mean "under `$HOME/projects`".
fn nest_root() -> Option<PathBuf> {
    crate::managed_agents::nest_dir()
}

/// `$HOME/projects`, where the reports these links point at are written.
fn projects_root() -> Option<PathBuf> {
    dirs::home_dir().map(|home| home.join("projects"))
}

/// The canonical roots a candidate from `sender_pubkey` may resolve inside.
///
/// A root that does not exist, or cannot be canonicalized, is dropped: on a
/// machine with neither `~/.buzz` nor `$HOME/projects` the list is empty and
/// every candidate is "not a link", which is the correct answer rather than
/// an error. `$HOME` itself is never a root.
pub(super) fn path_link_roots(sender_pubkey: Option<&str>) -> Vec<PathBuf> {
    let mut roots = Vec::new();
    for root in sender_workdir_root(sender_pubkey)
        .into_iter()
        .chain(nest_root())
        .chain(projects_root())
    {
        if let Ok(canonical) = root.canonicalize() {
            if canonical.is_dir() && !roots.contains(&canonical) {
                roots.push(canonical);
            }
        }
    }
    roots
}

/// Refuse a sender pubkey longer than the DTO admits.
fn check_sender_pubkey(sender_pubkey: Option<&str>) -> Result<(), String> {
    match sender_pubkey {
        Some(pubkey) if pubkey.len() > MAX_SENDER_PUBKEY_BYTES => Err(format!(
            "sender pubkey exceeds {MAX_SENDER_PUBKEY_BYTES} bytes"
        )),
        _ => Ok(()),
    }
}

fn extension_lowercase(path: &Path) -> Option<String> {
    path.extension()
        .and_then(|extension| extension.to_str())
        .map(|extension| extension.to_ascii_lowercase())
}

fn markdown_extension(path: &Path) -> bool {
    extension_lowercase(path)
        .map(|extension| MARKDOWN_EXTENSIONS.contains(&extension.as_str()))
        .unwrap_or(false)
}

/// True when the file's extension is one the OS handler treats as a document.
fn openable_extension(path: &Path) -> bool {
    extension_lowercase(path)
        .map(|extension| OPENABLE_EXTENSIONS.contains(&extension.as_str()))
        .unwrap_or(false)
}

/// True when the file carries any executable bit.
///
/// The second half of the "what does the handler do with it" guard: a
/// `report.md` that is also `chmod +x` is an anomaly for a file a *message*
/// named, and refusing it costs a real document nothing. Windows has no such
/// bit, so there it is always false and the extension allowlist stands alone.
#[cfg(unix)]
fn is_executable(metadata: &std::fs::Metadata) -> bool {
    use std::os::unix::fs::PermissionsExt;
    metadata.permissions().mode() & 0o111 != 0
}

#[cfg(not(unix))]
fn is_executable(_metadata: &std::fs::Metadata) -> bool {
    false
}

/// Refuse, by inspection alone, a candidate whose *lexical* form can name
/// something outside the roots or off this machine.
///
/// This runs before `canonicalize`, because `canonicalize` is itself the
/// dangerous call: `Path::is_absolute` accepts a Windows UNC path, and
/// canonicalizing `\\attacker\share\note.md` opens an SMB connection and an
/// NTLM handshake to a host named by relay-sourced text — credential exposure
/// driven by a message, before any containment logic has run. The shapes that
/// can do that are therefore rejected without touching a filesystem:
///
/// - a backslash anywhere: UNC (`\\host\share`), verbatim (`\\?\`), device
///   (`\\.\`) and every Windows spelling of a drive-relative path;
/// - a leading `//`, the forward-slash spelling of a UNC share;
/// - a `Prefix` or `RootDir` component on a candidate that is not absolute —
///   `C:file` is *relative* on Windows yet replaces the whole left-hand side
///   of a `join`;
/// - a `..` component, which a join would resolve outside the root.
pub(super) fn is_lexically_refused(candidate: &str) -> bool {
    if candidate.contains('\\') || candidate.starts_with("//") {
        return true;
    }
    let path = Path::new(candidate);
    let absolute = path.is_absolute();
    path.components().any(|component| match component {
        Component::ParentDir => true,
        Component::Prefix(_) | Component::RootDir => !absolute,
        _ => false,
    })
}

/// Resolve `candidate` against already-canonical `roots`.
///
/// `Err` means the candidate was refused outright (over the byte cap).
/// `Ok(None)` means "not a link" — a missing file, a directory, a path
/// outside every root, a symlink whose target leaves its root, or a file the
/// OS handler would run rather than display. `Ok(Some(_))` is an inert,
/// non-executable regular file inside a root, named by its canonical path.
pub(super) fn resolve_within_roots(
    candidate: &str,
    roots: &[PathBuf],
) -> Result<Option<PathLinkTarget>, String> {
    // Enforced before any filesystem call: an over-length candidate never
    // reaches `canonicalize`.
    if candidate.len() > MAX_PATH_LINK_BYTES {
        return Err(format!(
            "path link candidate exceeds {MAX_PATH_LINK_BYTES} bytes"
        ));
    }
    if candidate.is_empty() || candidate.contains('\0') {
        return Ok(None);
    }
    // Lexical refusal comes first, so no network or device path is ever
    // handed to `canonicalize`.
    if is_lexically_refused(candidate) {
        return Ok(None);
    }

    let candidate_path = Path::new(candidate);
    let absolute = candidate_path.is_absolute();
    for root in roots {
        let joined = if absolute {
            // Lexical containment, again before any filesystem call: an
            // absolute candidate that is not spelled under this root is not
            // canonicalized against it at all.
            if !candidate_path.starts_with(root) {
                continue;
            }
            candidate_path.to_path_buf()
        } else {
            root.join(candidate)
        };
        // Canonicalizing resolves every symlink, so a candidate that leaves
        // its root — however it left — fails the prefix check below.
        let Ok(canonical) = joined.canonicalize() else {
            continue;
        };
        // Component-wise: `/a/bc` does not start with `/a/b`, so the prefix is
        // always a whole path segment.
        if !canonical.starts_with(root) {
            continue;
        }
        let Ok(metadata) = canonical.metadata() else {
            continue;
        };
        if !metadata.is_file() {
            continue;
        }
        // Containment decided *which* file. These two decide what handing it
        // to the OS default handler can do.
        if !openable_extension(&canonical) || is_executable(&metadata) {
            continue;
        }
        // A canonical path that is not UTF-8 could not survive the round trip
        // back through the webview, so it is not offered as a link.
        let (Some(path), Some(filename)) = (
            canonical.to_str().map(str::to_owned),
            canonical
                .file_name()
                .and_then(|name| name.to_str())
                .map(str::to_owned),
        ) else {
            continue;
        };
        let size_bytes = metadata.len();
        let kind = if markdown_extension(&canonical) && size_bytes <= MAX_PATH_LINK_MARKDOWN_BYTES {
            PathLinkKind::Markdown
        } else {
            PathLinkKind::File
        };
        return Ok(Some(PathLinkTarget {
            path,
            filename,
            kind,
            size_bytes,
        }));
    }
    Ok(None)
}

/// Read at most `limit` bytes from `path`.
///
/// The bound is on the bytes this call actually reads, not on a size some
/// earlier call looked up: a file that grew in between still costs at most
/// `limit` bytes of memory here.
pub(super) fn read_bounded(path: &Path, limit: u64) -> Result<Vec<u8>, String> {
    let file =
        std::fs::File::open(path).map_err(|error| format!("read {}: {error}", path.display()))?;
    let mut bytes = Vec::new();
    file.take(limit)
        .read_to_end(&mut bytes)
        .map_err(|error| format!("read {}: {error}", path.display()))?;
    Ok(bytes)
}

/// Read a resolved markdown document, bounded by the bytes actually read.
///
/// The file is re-opened here rather than trusting the size recorded at
/// resolve time, and one byte past the cap is read so a document that grew in
/// between is refused instead of silently truncated.
pub(super) fn read_markdown_within_cap(path: &Path) -> Result<String, String> {
    let bytes = read_bounded(path, MAX_PATH_LINK_MARKDOWN_BYTES + 1)?;
    if bytes.len() as u64 > MAX_PATH_LINK_MARKDOWN_BYTES {
        return Err("This file is too large to preview.".to_string());
    }
    String::from_utf8(bytes)
        .map_err(|_| "This file isn't valid text, so it can't be previewed.".to_string())
}

pub(super) fn resolve_blocking(
    candidate: &str,
    sender_pubkey: Option<&str>,
) -> Result<Option<PathLinkTarget>, String> {
    check_sender_pubkey(sender_pubkey)?;
    resolve_within_roots(candidate, &path_link_roots(sender_pubkey))
}

fn require_target(target: Option<PathLinkTarget>) -> Result<PathLinkTarget, String> {
    target.ok_or_else(|| "That file is no longer on this Mac.".to_string())
}

/// Re-resolve `candidate` against `roots` and hand the *resolved* target to
/// `open`.
///
/// The webview holds a resolved path across a hover, a message edit and a
/// click, so containment is re-proven here at the moment of opening rather
/// than trusted from whatever the webview kept. `open` is reached only for an
/// inert, non-executable regular file inside a root; `open_path_link` is this
/// function plus the OS handler, so the re-resolution cannot be removed
/// without removing the call the tests drive.
pub(super) fn open_resolved_path_link<F>(
    candidate: &str,
    sender_pubkey: Option<&str>,
    roots: &[PathBuf],
    open: F,
) -> Result<(), String>
where
    F: FnOnce(&PathLinkTarget) -> Result<(), String>,
{
    check_sender_pubkey(sender_pubkey)?;
    let target = require_target(resolve_within_roots(candidate, roots)?)?;
    open(&target)
}

/// Re-resolve `candidate` against `roots` and read it as a viewer document.
///
/// Refuses anything the resolver did not classify as a viewer-sized markdown
/// document, so the panel cannot be pointed at a PDF or an oversized file.
pub(super) fn read_markdown_path_link(
    candidate: &str,
    sender_pubkey: Option<&str>,
    roots: &[PathBuf],
) -> Result<String, String> {
    check_sender_pubkey(sender_pubkey)?;
    let target = require_target(resolve_within_roots(candidate, roots)?)?;
    if target.kind != PathLinkKind::Markdown {
        return Err("That file is not a markdown document the viewer can render.".to_string());
    }
    read_markdown_within_cap(Path::new(&target.path))
}

/// Resolve one inline-code token to a local file, or report that it is not a
/// link.
///
/// Called on hover or click only. `Ok(None)` is the ordinary answer for a
/// token that names nothing openable and renders as plain text; an `Err` is a
/// refusal the caller surfaces.
#[tauri::command]
pub async fn resolve_path_link(
    candidate: String,
    sender_pubkey: Option<String>,
) -> Result<Option<PathLinkTarget>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        resolve_blocking(&candidate, sender_pubkey.as_deref())
    })
    .await
    .map_err(|error| format!("path link resolve task failed: {error}"))?
}

/// Open a resolved path link with the OS default handler.
///
/// The candidate is resolved again here rather than trusting a path the
/// webview held on to, so containment and the openable-type check are
/// re-proven at the moment the file is opened. The file is handed to the
/// opener, never executed.
#[tauri::command]
pub async fn open_path_link(
    candidate: String,
    sender_pubkey: Option<String>,
    app: AppHandle,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        let sender_pubkey = sender_pubkey.as_deref();
        let roots = path_link_roots(sender_pubkey);
        open_resolved_path_link(&candidate, sender_pubkey, &roots, |target| {
            app.opener()
                .open_path(&target.path, None::<&str>)
                .map_err(|error| format!("open {}: {error}", target.filename))
        })
    })
    .await
    .map_err(|error| format!("path link open task failed: {error}"))?
}

/// Read a local markdown document for the in-app viewer panel.
///
/// Re-resolves the candidate (same containment check as `resolve_path_link`),
/// refuses anything the resolver did not classify as a viewer-sized markdown
/// document, and bounds the read at [`MAX_PATH_LINK_MARKDOWN_BYTES`].
#[tauri::command]
pub async fn read_path_link_markdown(
    candidate: String,
    sender_pubkey: Option<String>,
) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let sender_pubkey = sender_pubkey.as_deref();
        let roots = path_link_roots(sender_pubkey);
        read_markdown_path_link(&candidate, sender_pubkey, &roots)
    })
    .await
    .map_err(|error| format!("path link read task failed: {error}"))?
}
